fix: Configuration option typo

bharat opened the pull request:
There is a typo in the configuration option key: it is currently written as "storgage-id" instead of "storage-id." Both versions of the option will be supported as a transition until the end of this year. After that, the deprecated "storgage-id" option will be removed. Therefore, all users are strongly encouraged to update their configurations to use the correct "storage-id" key as soon as possible.

bharat committed:
Update README.md

Fix a minor typo

MelleD commented:
Hey @bharat,

thanks for the PR, but into the code it's also the same typo see:
https://github.com/MelleD/lovelace-expander-card/blob/main/src/configtype.ts#L40
and
https://github.com/MelleD/lovelace-expander-card/blob/main/src/ExpanderCard.svelte#L54

So you have to fix this as well and then this PR is breaking for the people you uses currently the id with typo

bharat commented:
Ah, I didn't dig deeper into this. But it makes sense that the docs are generated from the code so the typo lives in the code as well. What do you suggest?

I could go with a deprecation strategy, which at a surface level inspection would look like adding 'storage-id' as an additional optional string to the ExpanderConfig interface, then in svelte we could  do something like this:

`const configId = config['storage-id'] ?? config['storgage-id'];`

Then somewhere down the road, we could phase out support for the typo if you ever do an API-breaking major release. At least it wouldn't be propagating...

I'm happy to explore this route and test it out locally if you think it is a good approach.

MelleD commented:
You're tackling some issues. Let me try to summarize everything.

You're right, it would be better if the documentation or the table were generated from the file, but unfortunately, that's not the case right now. It would be good to create a new PR or feature request for that "issue".

Regarding the current problem, you're absolutely right about your approach, and it's user-friendly, but also more complex. Feel free to try it out, if you want :-).

However, I think for our case, it's enough to simply change everything or all places with the typo. Label it "breaking-change" in the PR and make a major version as a release.

I think that many people don't use the feature directly, and the fix would be easy and nothing break really "hard".

What do you think?

bharat committed:
Fix 'storgage' typo everywhere.

Try to provide a little backwards compatibility for those who were
using the incorrect spelling, but keep it out of the docs and the
interface so that we don't propagate the mistake.

bharat committed:
Merge branch 'main' into patch-1

bharat commented:
That makes sense. I see two options:

1. Just rename it all, label it "breaking-change" and let the chips fall where they may.

2. Rename it in the interface and the README, but in ExpanderCard.svelte put in a fallback. That way we won't propagate the typo directly, but it won't immediately break any existing installs. 

I'm leaning towards #2 because I feel bad for anybody who's got an install out there who's going to get silently or weirdly broken by a UI update. I adjusted this patch accordingly. That said, I defer to you and if you prefer #1 then I'll adjust it that way.

What's your preferred approach to testing? I'm about a week into HA over here so while I've played around with it quite a bit, this is my first foray into developing a custom mod. Many thanks for your hard work on this card- I use it quite a bit!

MelleD commented:
If you feel better then we go with option 2.

My proposal is, we make a bugfix release with both option this year and end of the year we remove the "old" typo option with a breaking change release. 
We describe this in the README that everybody should update the option till end of the year.

For testing I or you can build it locally on mine/your machine and then install it.

MelleD commented on src/configtype.ts:
```diff
@@ -37,7 +37,7 @@ export interface ExpanderConfig {
     'max-width-expanded'?: number;
     icon?: string;
-    'storgage-id'?: string;
+    'storage-id'?: string;
```
If we go with both option we need both optional parameter here

MelleD commented on README.md:
```diff
@@ -69,7 +69,7 @@ Yaml Options:
 | title-card-clickable      | boolean  | _false_       | true\|false            | Should the complete diff clickable?                   |
 | overlay-margin            | string   | _0.0em_       | css-size               | Margin from top right of expander button (if overlay) |
-| storgage-id               | string   | **optional**  | *                      | Save last expander state in local browser storage     |
+| storage-id                | string   | **optional**  | *                      | Save last expander state in local browser storage     |
```
Please add a section when we will remove the 'storgage-id' as option

bharat committed:
Add 'storgage-id' back to ExpanderConfig, and a deprecation warning in the README.

bharat commented:
Alrighty, I've been enjoying learning how these modules are put together. Thanks for bearing with me. I've added `storgage-id` back to ExpanderConfig and added a deprecation warning to the README. Not sure where is a great place to add this, so I put it below the table for now. Happy to move that around.

I tested it with the following YAML:

```
type: vertical-stack
cards:
  - type: custom:expander-card
    title: separate (storage)
    storage-id: expander-storage-id
    cards:
      - type: entities
        entities:
          - automation.alarm_away
  - type: custom:expander-card
    title: separate (storgage)
    storgage-id: expander-storgage-id
    cards:
      - type: entities
        entities:
          - automation.alarm_away
  - type: custom:expander-card
    title: combined (storage)
    storage-id: expander-combined
    cards:
      - type: entities
        entities:
          - automation.alarm_away
  - type: custom:expander-card
    title: combined (storgage)
    storgage-id: expander-combined
    cards:
      - type: entities
        entities:
          - automation.alarm_away
```

I verified in the console that all three of the keys are set, and that their values are stable and persist across reloads.

```
> Object.keys(localStorage).filter(k => k.includes('expander'))
(3) ['expander-open-expander-storgage-id', 'expander-open-expander-combined', 'expander-open-expander-storage-id']
```

Let me know what else I can do here!

MelleD commented:
Great thank you. Looks good for me
